lib/client/kubesession.go

```diff
@@ -0,0 +1,221 @@
+		return nil, trace.Wrap(err)
+	}
+
+	stream, err := streamproto.NewSessionStream(ws, streamproto.ClientHandshake{Mode: mode})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	term, err := terminal.New(tc.Stdin, tc.Stdout, tc.Stderr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	closeWait.Add(1)
+	go func() {
+		<-ctx.Done()
+		term.Close()
+		closeWait.Done()
+	}()
+
+	if term.IsAttached() {
+		// Put the terminal into raw mode. Note that this must be done before
+		// pipeInOut() as it may replace streams.
+		term.InitRaw(true)
+	}
+
+	handleOutgoingResizeEvent(ctx, stream, term)
+	handleIncomingResizeEvent(stream, closeWait, term)
+	s := &KubeSession{stream: stream, term: term, ctx: ctx, cancelFunc: cancel, closeWait: closeWait, meta: meta}
+	err = s.handleMFA(ctx, tc, mode)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s.pipeInOut()
+	return s, nil
+}
+
+func (s *KubeSession) cancel() {
+	s.cancelOnce.Do(func() {
+		s.cancelFunc()
+	})
+}
+
+func handleOutgoingResizeEvent(ctx context.Context, stream *streamproto.SessionStream, term *terminal.Terminal) {
+	go func() {
+		queue := stream.ResizeQueue()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case size := <-queue:
+				if size == nil {
+					return
+				}
+
+				term.Resize(int16(size.Width), int16(size.Height))
+			}
+		}
+	}()
+}
+
+func handleIncomingResizeEvent(stream *streamproto.SessionStream, closeWait *sync.WaitGroup, term *terminal.Terminal) error {
+	closeWait.Add(1)
+	go func() {
+		defer closeWait.Done()
+		events := term.Subscribe()
+
+		for {
+			event, more := <-events
+			_, ok := event.(terminal.ResizeEvent)
+			if ok {
+				w, h, err := term.Size()
+				if err != nil {
+					return
+				}
+
+				size := remotecommand.TerminalSize{Width: uint16(w), Height: uint16(h)}
+				err = stream.Resize(&size)
+				if err != nil {
+					fmt.Printf("Error attempting to resize terminal: %v\n\r", err)
+				}
+			}
+
+			if !more {
+				break
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (s *KubeSession) handleMFA(ctx context.Context, tc *TeleportClient, mode types.SessionParticipantMode) error {
+	if s.stream.MFARequired && mode == types.SessionModeratorMode {
+		proxy, err := tc.ConnectToProxy(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		auth, err := proxy.ConnectToCluster(ctx, s.meta.GetClustername(), false)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		subCtx, cancel := context.WithCancel(ctx)
+		go func() {
+			<-ctx.Done()
+			cancel()
+		}()
+
+		go runPresenceTask(subCtx, s.term.Stdout(), auth, tc, s.meta.GetSessionID())
+	}
+
+	return nil
+}
+
+// pipeInOut starts background tasks that copy input to and from the terminal.
+func (s *KubeSession) pipeInOut() {
+	go func() {
+		defer s.cancel()
+		_, err := io.Copy(s.term.Stdout(), s.stream)
+		if err != nil {
+			fmt.Printf("Error while reading remote stream: %v\n\r", err.Error())
```

Comment:
I don't think a logging lib is suitable here. They usually do input manipulation like adding newlines etc which won't work since we run in raw terminal mode. To work around this I've switched to `fmt.Fprint` and a writer wrapper that synchronizes writes.